### PR TITLE
Update documentation and example config for HTTPS by default.

### PR DESCRIPTION
**Pull request by jpwhite4**
I'm going to add some faq entries to head off any common error messages seen with the new defaults.

### CODE REVIEW

**Comment by jtpalmer** [comment]
If we're going to default to a more production-like configuration then I'd just as well drop the use of port 8080 and use 443 everywhere.

File: docs/configuration.md [hunk 1]
```diff
@@ -120,56 +120,68 @@ See the [Hierarchy Guide](hierarchy.html) for more details.
...
+        # Use HTTP Strict Transport Security to force client to use secure connections only
+        Header always set Strict-Transport-Security "max-age=31536000; includeSubDomains"
+
         DocumentRoot /usr/share/xdmod/html
+
         <Directory /usr/share/xdmod/html>
             Options FollowSymLinks
             AllowOverride All
-            DirectoryIndex index.php
+            DirectoryIndex index.php index.html
```
Comment (jtpalmer) [inline docs/configuration.md:168]: Why is this changed?

File: tests/ci/bootstrap.sh [hunk 2]
```diff
@@ -28,6 +32,7 @@ then
 
     rm -rf /var/lib/mysql && mkdir -p /var/lib/mysql
     yum -y install ~/rpmbuild/RPMS/*/*.rpm
+    switch_off_ssl
```
Comment (plessbd) [inline tests/ci/bootstrap.sh:35]: looks like you might need this in an upgrade as well...

File: docs/configuration.md (hunk 1 above)
Comment (jpwhite4) [reply docs/configuration.md:168]: cut-and-pasta

**Comment by jpwhite4** [comment]
> If we're going to default to a more production-like configuration then I'd just as well drop the use of port 8080 and use 443 everywhere.

well the issue is that the redhat conf already ships with stuff listening on port 443:
image

the shipped config actually works out of the box on a new centos system as it uses the self-signed cert that comes with it.

File: configuration/apache.conf [hunk 3]
```diff
@@ -1,33 +1,43 @@
-# We recommend that you use HTTPS in production.  If your web server is
-# not already listening on port 443 you will need to uncomment the
-# following "Listen" directive or add it to the appropriate section in
-# your Apache configuration:
+# Template Apache configuration file for XDMoD. This file should
+# be edited to specify the correct site-specific settings:
 #
-#Listen 443
+# The ServerName setting should be updated.
```
[the remark this answers is not in the document]
Comment (jpwhite4) [reply configuration/apache.conf:4]: The design is to (a) summarize the necessary changes at the top and (b) annotate each required change where it needs to be. I think this strategy will lead to fewer support requests due to people not reading the docs.

File: templates/apache.conf [hunk 4]
```diff
@@ -0,0 +1,84 @@
+# TEMPLATE Apache configuration file for XDMoD. This file should
```
Comment (jtpalmer) [inline templates/apache.conf:1]:
```suggestion
# TEMPLATE Apache configuration file for Open XDMoD. This file should
```

File: templates/apache.conf [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+# be copied to the apache configuration directory and
```
Comment (jtpalmer) [inline templates/apache.conf:2]:
```suggestion
# be copied to the Apache configuration directory and 
```

File: templates/apache.conf [hunk 6]
```diff
@@ -0,0 +1,84 @@
+# TEMPLATE Apache configuration file for XDMoD. This file should
+# be copied to the apache configuration directory and 
+# edited to specify the correct site-specific settings.
+#
+# On Centos 7 and RHEL 7, this file should be copied
```
Comment (jtpalmer) [inline templates/apache.conf:5]:
```suggestion
# On CentOS 7 and RHEL 7, this file should be copied
```

File: templates/apache.conf [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+#
+# The SSLCertificateFile and SSLCertificateKeyFile settings should
+# be updated to specify paths to the valid SSL certificates for the
+# site.
+#
+# Optionally the port number in the VirtualHost section can be updated
+# from 443 to the desired listen port.
+#
+# The server name and port number in the Apache configuration must match
+# the site_address and user_manual settings in the XDMoD portal_settings.ini
```
Comment (jtpalmer) [inline templates/apache.conf:24]:
```suggestion
# the site_address and user_manual settings in the Open XDMoD portal_settings.ini
```

File: templates/apache.conf [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+# If the server is not already configured to listen on port 443 then the
+# following Listen command should be uncommented.
+#Listen 443
+
+<VirtualHost *:443>
+    # The ServerName and ServerAdmin parameters should be updated.
+    ServerName localhost
+    ServerAdmin postmaster@localhost
+
+    # Production XDMoD instances should use HTTPS
```
Comment (jtpalmer) [inline templates/apache.conf:37]:
```suggestion
    # Production Open XDMoD instances should use HTTPS
```

File: templates/apache.conf [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+    Header always set Strict-Transport-Security "max-age=31536000; includeSubDomains"
+
+    DocumentRoot /usr/share/xdmod/html
+
+    <Directory /usr/share/xdmod/html>
+        Options FollowSymLinks
+        AllowOverride All
+        DirectoryIndex index.php
+
+        # Apache 2.4 access controls.
```
Comment (jtpalmer) [inline templates/apache.conf:59]: We started requiring Apache 2.4 so we might want to remove this if we don't support Apache 2.2.  Not necessarily as part of these changes.

File: templates/apache.conf [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+        RewriteRule (.*) index.php [L]
+    </Directory>
+
+    ## SimpleSAML Single Sign On authentication.
+    #SetEnv SIMPLESAMLPHP_CONFIG_DIR /etc/xdmod/simplesamlphp/config
+    #Alias /simplesaml /usr/share/xdmod/vendor/simplesamlphp/simplesamlphp/www
+    #<Directory /usr/share/xdmod/vendor/simplesamlphp/simplesamlphp/www>
+    #    Options FollowSymLinks
+    #    AllowOverride All
+    #    # Apache 2.4 access controls.
```
Comment (jtpalmer) [inline templates/apache.conf:76]: Another Apache 2.4 comment that isn't really necessary anymore.

File: docs/configuration.md [hunk 11]
```diff
@@ -120,56 +120,85 @@ See the [Hierarchy Guide](hierarchy.html) for more details.
...
-            <IfModule mod_authz_core.c>
-                Require all granted
-            </IfModule>
-        </Directory>
-        <Directory /usr/share/xdmod/html/rest>
-            RewriteEngine On
-            RewriteRule (.*) index.php [L]
-        </Directory>
-    </VirtualHost>
+For Centos 7 and RHEL 7 the template file should be copied to `/etc/httpd/conf.d/xdmod.conf`
```
Comment (jtpalmer) [inline docs/configuration.md:128]:
```suggestion
For CentOS 7 and RHEL 7 the template file should be copied to `/etc/httpd/conf.d/xdmod.conf`
```

File: docs/configuration.md [hunk 12]
```diff
@@ -120,56 +120,85 @@ See the [Hierarchy Guide](hierarchy.html) for more details.
...
+for SSL configuration information.
 
-We recommend that you use HTTPS in production.  This will require
-additional configuration.
+The name and port of the server must match with the `site_address` and `user_manual`
+configuration settings in `portal_settings.ini`.
 
-    Listen 443
+The template configuration file also configures the webserver to send the `Strict-Transport-Security` HTTP Header
+to indicate to  web browsers that the XDMoD instance should only be accessed using HTTPS.
```
Comment (jtpalmer) [inline docs/configuration.md:145]:
```suggestion
to indicate to  web browsers that the Open XDMoD instance should only be accessed using HTTPS.
```

File: docs/configuration.md [hunk 13]
```diff
@@ -120,56 +120,85 @@ See the [Hierarchy Guide](hierarchy.html) for more details.
...
+
+```apache
     <VirtualHost *:443>
+        # The ServerName and ServerAdmin parameters should be updated.
+        ServerName localhost
+        ServerAdmin postmaster@localhost
 
-        # Customize this section using your SSL certificate.
+        # Production XDMoD instances should use HTTPS
```
Comment (jtpalmer) [inline docs/configuration.md:153]:
```suggestion
        # Production Open XDMoD instances should use HTTPS
```

File: docs/faq.md [hunk 14]
```diff
@@ -166,3 +166,15 @@ to recover the User to Acl relations.
 This indicates that the `-r` or `--recover` flag was present when calling `acl-config`
 but there were no backup tables found to recover from. Remove the `-r` or `--recover` flag
 and run `acl-config` again.
+
+### Why do I see the error message "Bad Request Your browser sent a request that this server could not understand." in the browser instead of the portal?
+
+This message is displayed if an HTTPS site is accessed via the HTTP protocol. The default
+example webserver configuration in XDMoD 9.0 and later is to enable HTTPS. HTTPS
```
Comment (jtpalmer) [inline docs/faq.md:173]:
```suggestion
example webserver configuration in Open XDMoD 9.0 and later is to enable HTTPS. HTTPS
```

File: docs/software-requirements.md [hunk 15]
```diff
@@ -105,10 +109,24 @@ by adding the following, but substituting your timezone:
 
 The PHP website contains the full list of supported [timezones][].
 
+Production XDMoD instances should use HTTPS, which is enabled via the webserver
```
Comment (jtpalmer) [inline docs/software-requirements.md:112]:
```suggestion
Production Open XDMoD instances should use HTTPS, which is enabled via the webserver
```

File: docs/upgrade.md [hunk 16]
```diff
@@ -129,6 +129,16 @@ necessary.
 The `xdmod-upgrade` script will migrate user editable configuration files to
 the new version.
 
+The example apache configuration file has changed to default to HTTPS.
+If the existing apache configuration file has been previously
+edited then it will not be replaced by an RPM upgrade.
+However, the RPM upgrade will replace the existing configuration file
+ with the new one if the existing file has not been modified from default.
+If you are upgrading an existing XDMoD installation that used the default
```
Comment (jtpalmer) [inline docs/upgrade.md:137]:
```suggestion
If you are upgrading an existing Open XDMoD installation that used the default
```

File: docs/upgrade.md [hunk 17]
```diff
@@ -129,6 +129,16 @@ necessary.
...
+apache configuration file (http port 8080) then it is recommended to
+switch to HTTPS and update the apache configuration with valid SSL certificates.
+XDMoD will still work with the previous HTTP only apache configuration.
```
Comment (jtpalmer) [inline docs/upgrade.md:140]:
```suggestion
Open XDMoD will still work with the previous HTTP only Apache configuration.
```

File: docs/configuration.md [hunk 18]
```diff
@@ -120,56 +120,85 @@ See the [Hierarchy Guide](hierarchy.html) for more details.
 Apache Configuration
 --------------------
 
-Uses port 8080 by default, if changed, must also be changed in
-`portal_settings.ini`.
+A template apache configuration file is provided. The path is `/usr/share/xdmod/templates/apache.conf`
```
Comment (jtpalmer) [inline docs/configuration.md:123]:
```suggestion
A template Apache configuration file is provided. The path is `/usr/share/xdmod/templates/apache.conf`
```

File: docs/faq.md [hunk 19]
```diff
@@ -166,3 +166,15 @@ to recover the User to Acl relations.
...
+sites should use the https:// prefix in the web address.
+
+### Error "SSLCertificateFile: file '/etc/pki/tls/certs/localhost.crt' does not exist or is empty" when trying to start the webserver
+
+The example template apache configuration file must be edited to specify the path to
```
Comment (jtpalmer) [inline docs/faq.md:178]:
```suggestion
The template Apache configuration file must be edited to specify the path to
```

File: docs/upgrade.md [hunk 20]
```diff
@@ -129,6 +129,16 @@ necessary.
 The `xdmod-upgrade` script will migrate user editable configuration files to
 the new version.
 
+The example apache configuration file has changed to default to HTTPS.
```
Comment (jtpalmer) [inline docs/upgrade.md:132]:
```suggestion
The Apache template configuration file has changed to default to HTTPS.
```

File: docs/upgrade.md [hunk 21]
```diff
@@ -129,6 +129,16 @@ necessary.
...
+If the existing apache configuration file has been previously
```
Comment (jtpalmer) [inline docs/upgrade.md:133]:
```suggestion
If the existing Apache configuration file has been previously
```

File: docs/upgrade.md [hunk 22]
```diff
@@ -129,6 +129,16 @@ necessary.
...
+edited then it will not be replaced by an RPM upgrade.
+However, the RPM upgrade will replace the existing configuration file
+ with the new one if the existing file has not been modified from default.
+If you are upgrading an existing XDMoD installation that used the default
+apache configuration file (http port 8080) then it is recommended to
```
Comment (jtpalmer) [inline docs/upgrade.md:138]:
```suggestion
Apache configuration file (http port 8080) then it is recommended to
```

File: docs/upgrade.md [hunk 23]
```diff
@@ -129,6 +129,16 @@ necessary.
...
+switch to HTTPS and update the apache configuration with valid SSL certificates.
```
Comment (jtpalmer) [inline docs/upgrade.md:139]:
```suggestion
switch to HTTPS and update the Apache configuration with valid SSL certificates.
```

**Comment by plessbd** [comment]
replaced by #1340

File: docs/configuration.md [hunk 24]
```diff
@@ -120,56 +120,84 @@ See the [Hierarchy Guide](hierarchy.html) for more details.
...
+        #    AllowOverride All
+        #    <IfModule mod_authz_core.c>
+        #        Require all granted
+        #    </IfModule>
+        #</Directory>
+
+        ErrorLog /var/log/xdmod/apache-error.log
+        CustomLog /var/log/xdmod/apache-access.log combined
     </VirtualHost>
+```
```
Comment (jtpalmer) [inline docs/configuration.md:200]: This shouldn't be indented anymore if it's in a `` ``` `` block.

File: docs/software-requirements.md [hunk 25]
```diff
@@ -105,10 +109,24 @@ by adding the following, but substituting your timezone:
 
 The PHP website contains the full list of supported [timezones][].
 
+Production Open XDMoD instances should use HTTPS, which is enabled via the webserver
+configuration (see below). Following a defense in depth approach to security,
+it is also recommended to set the php `session.cookie_secure` configuration.
+This ensures that php cookies are only sent over secure connections.
+
+```ini
+    session.cookie_secure = 1
```
Comment (jtpalmer) [inline docs/software-requirements.md:118]: This doesn't need to be indented.
